app/src/org/commcare/models/database/user/UserDatabaseUpgrader.java

```diff
@@ -648,6 +664,50 @@ private boolean upgradeTwentyThreeTwentyFour(SQLiteDatabase db) {
         }
     }
 
+    // check for integrity of SSD records and wipes the whole table in case of any discrepancy
+    private boolean upgradeTwentyFourTwentyFive(SQLiteDatabase db) {
+        db.beginTransaction();
+        try {
+            boolean strandedRecordObserved = false;
+            SqlStorage<FormRecord> formRecordStorage = UserDbUpgradeUtils.getFormRecordStorage(c, db, FormRecord.class);
+            SqlStorage<SessionStateDescriptor> ssdStorage = new SqlStorage<>(
+                    SessionStateDescriptor.STORAGE_KEY,
+                    SessionStateDescriptor.class,
+                    new ConcreteAndroidDbHelper(c, db));
+            for (SessionStateDescriptor ssd : ssdStorage) {
+                // we are in invalid state if formRecord with corresponding ssd form id
+                // either doesn't exist or has status unstarted
+                try {
+                    FormRecord formRecord = formRecordStorage.read(ssd.getFormRecordId());
+                    if (formRecord.getStatus().contentEquals(FormRecord.STATUS_UNSTARTED)) {
+                        strandedRecordObserved = true;
+                        break;
+                    }
+                } catch (NoSuchElementException e) {
+                    strandedRecordObserved = true;
+                    break;
+                }
+            }
+
+            if (strandedRecordObserved) {
+                SqlStorage.wipeTable(db, SessionStateDescriptor.STORAGE_KEY);
+
+                // Since we have wiped out SSD records, we won't be able to resume
+                // incomplete forms with their earlier session state. Therfore we are
+                // going to delete all incomplete form records as well
+                for (FormRecord formRecord : formRecordStorage) {
+                    if (formRecord.getStatus().contentEquals(FormRecord.STATUS_INCOMPLETE)) {
+                        formRecordStorage.remove(formRecord);
```

Comment:
So I generally think of deleting in-place like this (while still using an iterator) as being a dangerous operation, and would assume we'd need to make a "todelete" list and then apply the deletions afterwards. Am I incorrect about that WRT our iterators?